trl/data_utils.py

```diff
@@ -513,3 +513,88 @@ def maybe_convert_to_chatml(example: dict[str, list]) -> dict[str, list]:
         example["messages"] = example.pop("conversations")
 
     return example
+
+
+def find_largest_leq_k(ind_len_tuples, k):
+    valid = [(i, t[0]) for i, t in enumerate(ind_len_tuples) if t[0] <= k]
+    return max(valid, key=lambda x: x[1])[0] if valid else -1
+
+
+def greedy_knapsack(numbers: list[int], capacity: int) -> list[list]:
+    """
+    An efficient greedy algorithm with binary search for the knapsack problem.
+    Adapted from Llama Factory - https://github.com/hiyouga/LLaMA-Factory/blob/main/src/llamafactory/data/processor/processor_utils.py#L62
+    """
+    numbers = sorted(numbers, key=lambda x: x[0])  # sort numbers in ascending order for binary search
+    knapsacks = []
+
+    while numbers and (numbers[0][0] < capacity):
+        current_knapsack = []
+        remaining_capacity = capacity
+
+        while True:
+            index = find_largest_leq_k(numbers, remaining_capacity)
+            if index == -1:
+                break
+            remaining_capacity -= numbers[index][0]  # update the remaining capacity
+            current_knapsack.append(numbers.pop(index)[1])  # add the number to knapsack
+
+        knapsacks.append(current_knapsack)
+
+    return knapsacks, [[tup[1]] for tup in numbers]
+
+
+def pack_examples_smarter(examples: dict[str, list[list]], seq_length: int) -> dict[str, list[list]]:
```

Comment:
wdyt of
```suggestion
def pack_examples_knapsack(examples: dict[str, list[list]], seq_length: int) -> dict[str, list[list]]:
```


or even

```suggestion
def pack_examples_knapsack_greedy(examples: dict[str, list[list]], seq_length: int) -> dict[str, list[list]]:
```

Reply:
fixed!